### PR TITLE
[scudo] allocation_ring_buffer_size <= 0 disables buffer

### PR DESCRIPTION
**Pull request by fmayer**
Prevent a null pointer exception for allocation_ring_buffer_size < 0.

### CODE REVIEW

**Review by hctim — COMMENTED** [review]
`compiler-rt/lib/scudo/standalone/combined.h:1494`

Also this (to avoid mapping it as well)?

```
1492   void mapAndInitializeRingBuffer() {
1493     u32 AllocationRingBufferSize =
1494         static_cast<u32>(getFlags()->allocation_ring_buffer_size);
////     Hoist this check up, and make it be |getFlags()->allocation_ring_buffer_size < 1| to avoid the
////     type punning?
1495     if (AllocationRingBufferSize < 1)
1496       return;
1497     MemMapT MemMap;
1498     MemMap.map(
1499         /*Addr=*/0U,
1500         roundUp(ringBufferSizeInBytes(AllocationRingBufferSize),
```

**Comment by fmayer** [comment]
> `compiler-rt/lib/scudo/standalone/combined.h:1494`
> 
> Also this (to avoid mapping it as well)?
> 
> ```
> 1492   void mapAndInitializeRingBuffer() {
> 1493     u32 AllocationRingBufferSize =
> 1494         static_cast<u32>(getFlags()->allocation_ring_buffer_size);
> ////     Hoist this check up, and make it be |getFlags()->allocation_ring_buffer_size < 1| to avoid the
> ////     type punning?
> 1495     if (AllocationRingBufferSize < 1)
> 1496       return;
> 1497     MemMapT MemMap;
> 1498     MemMap.map(
> 1499         /*Addr=*/0U,
> 1500         roundUp(ringBufferSizeInBytes(AllocationRingBufferSize),
> ```

ah, yes. thanks

**Review by hctim — APPROVED** [review]
LGTM

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 1]
```diff
@@ -886,7 +886,7 @@ class Allocator {
 
   void setTrackAllocationStacks(bool Track) {
     initThreadMaybe();
-    if (getFlags()->allocation_ring_buffer_size == 0) {
+    if (getFlags()->allocation_ring_buffer_size < 1) {
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/combined.h:889]: Can we do this with `<= 0`? I think the intention will be more clear

In addition, if we can make `allocation_ring_buffer_size` unsigned, `== 0` will be better

Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h]: > Can we do this with <= 0? I think the intention will be more clear

OK

> In addition, if we can make allocation_ring_buffer_size unsigned, == 0 will be better

I would rather not change the parser for this. There is precedence to making negative values disable, for `release_to_os_interval_ms`.

**Comment by fmayer** [comment]
PTAL

**Review by cferris1000 — APPROVED** [review]

**Pull request merged**